### ISSUE TITLE
deepRAM

### ISSUE THREAD

**Issue by iwasakishuto** [iwasakishuto/Papers-With-Comments#9]
# deepRAM
- Comprehensive evaluation of deep learning architectures for prediction of DNA/RNA sequence binding specificities.
- Submitted on Tue, 29 Jan 2019 (v1)

### Link
- [paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6612801/)
- [Github](https://github.com/MedChaabane/deepRAM)

### Summary
- DNA- and RNA-binding proteins are involve in many biological processes including **transcription**, **translation** and **alternative splicing**.
- Present an **Automatic** exploration of deep learning architectures **for predicting DNA- and RNA-binding specificity.**
- Deeper more complex architectures like hybrid CNN/RNN models **outperform other methods in terms of accuracy**, but that requires the availability of sufficient training data, and these network are harder to interpret. Therefore **their usefulness in motif discovery might be limited.**

### Model Architectures

![Fig.1.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6612801/bin/btz339f1.jpg)
<center><small>Fig.1. Overviw of the deep learning architectures evaluated in this paper</small></center>

#### "one-hot encoding" vs "_k_-mer embedding"
These are the ways of representing the input sequence.

- one-hot encoding
    - Input sequence is represented by a 4×L matrix.
        - L is the length of the sequence.
        - Each position in the sequence is associated with a non-zero vector of length 4.
- _k_-mer embedding
    - Split the sequence into overlapping _k_-mers of length _k_ using a sliding window with stride _s_. (in Fig.1. _s=2_, _d=3_)
    - Map each _k_-mer into _d_-dimensional vector space using the **word2vec** algorithm.
        - Unsupervised learning algorithm which maps _k_-mers **from the vocabulary to vectors of real numbers in a low-dimensional space.**
        - The embedding representation of _k_-mer is computed in such a way that **their context is preserved.**

#### "CNN" vs "RNN" vs "CNN/RNN"
- CNN
    - **1D convolution**: Each convolutional filter is an M×N matrix with M being the window size and N being the number of input channels. (For the first conv layer, N equals the input representation dimension; 4 for one-hot, _d_ for _k_-mer.)
    - **The first convolutional layer can be thought of as a motif detector** where each filter is analogous to a **PWM(Position weight matrices)**
- RNN
    - GRU units and LSTM units: Using a gating mechanism to capture short- and long-term dependencies in sequences.
    - bi-RNA (forward and backward)
- Hybrid models
    - The convolutional stage is followed by an RNN stage.

#### Existing variants

|Name|Input|Model|
|:-:|:-|:-|
|[DeepBind](https://www.nature.com/articles/nbt.3300)|one-hot|single CNN|
|[DanQ](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4914104/)|one-hot|single layer CNN followed by a bidirectional LSTM|
|[KEGRU](https://www.nature.com/articles/s41598-018-33321-1)|_k_-mer|GRU units and dilated multi-layer CNN|

#### Additional variants

|Name|Input|Model|
|:-:|:-|:-|
|DeepBind*|one-hot|multi-layer CNN|
|DanQ*|one-hot|DanQ with multiple layers of convolution|
|DeepBind-E*|_k_-mer|multi-layer CNN|
|ECLSTM|_k_-mer|single layer CNN and LSTM|
|ECBLSTM|_k_-mer|single layer CNN and bidirectional LSTM|

#### Model Selection
This model selection strategy is based on the one used by the authors of [DeepBind](https://www.nature.com/articles/nbt.3300).

- Hyper Parameter Tuning
    - For each dataset, **sampling 40 hyper-parameter settings** from all possible combinations
    - Each parameter setting is evaluated using its **area under the ROC curve (AUC)** in **3-fold cross-validation.**
- Select Model
    - Use the selected best hyper-parameter setting to train 5 models **(to avoid the effect of random initialization)** using the full training data.
    - Choose the model with the best training performance as the final selected.

#### Training
- Minimize the cross-entropy objective function.
- Optimize by SGD or Adagrad. (the choice is made as part of the model selection process.)
- Batch size is 128.
- Use multiple regularization schemes: dropout, weight decay, and early stopping.

#### Motif extraction
As I mentioned before, **extract motifs from the first convolutional layer.**

1. Feed all test sequence through the convolution stage.
2. For each filter, **extract all sequence fragments that activate the filter** and **use only activations that are greater than half of the filter's maximum value.**
3. Once all the sequence fragments are extracted, they are stacked and the nucleotide frequencies are counted to form a **PFM(Position Frequency Matrix)**.
4. Sequence logos are then constructed using [WebLogo](https://weblogo.berkeley.edu/).
5. The discovered motifs are aligned using [TOMTOM](http://meme-suite.org/tools/tomtom) against known motifs from [CISBP-RNA](http://cisbp-rna.ccbr.utoronto.ca/) for RBPs and [JASPAR](http://jaspar.genereg.net/) for transcription factors.

### Results
1. Deeper is better with sufficient training data.
2. _k_-mer embedding boosts model performance.
3. Dilated convolution is a valuable architecture parameter to consider.
4. There is negligible difference between GRU and LSTM.
5. RNNs lead to a representation which has reduced interpretability compared to that of CNNs.

<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6612801/figure/btz339-F2/"><img src="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6612801/bin/btz339f2.jpg" alt="Fig.2"></a>
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6612801/figure/btz339-F3/"><img src="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6612801/bin/btz339f3.jpg" alt="Fig.3"></a>

### DataSets
- [Encyclopedia of DNA Elements (ENCODE) project (ENCODE-Project-Consortium, 2012)](https://www.encodeproject.org/)
- verified binding site of RBPs derived from [large-scale CLIP-seq experiments.](https://www.ncbi.nlm.nih.gov/pubmed/26787667/)
